lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,126 @@
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "cql3/functions/function_name.hh"
+#include "utils/overloaded_functor.hh"
+#include "cql3/prepared_statements_cache.hh"
+#include "wasmtime.hh"
+#include "lang/wasm.hh"
+#include <exception>
+#include <seastar/core/metrics.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/scheduling.hh>
+#include <seastar/core/shared_mutex.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/defer.hh>
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <queue>
+#include <unordered_map>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct wasm_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class instance_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+private:
+    stats _stats;
+    seastar::metrics::metric_groups _metrics;
+
+public:
+    stats& shard_stats();
+
+    void setup_metrics();
+private:
+    using cache_key_type = cql3::functions::function_name;
+
+    struct lru_entry_type;
+
+    struct cache_entry_type {
+        seastar::scheduling_group scheduling_group;
+        std::vector<data_type> arg_types;
+        seastar::shared_mutex mutex;
+        std::optional<wasm_instance> instance;
+        // iterator points to _lru.end() when the entry is being used (at that point, it is not in lru)
+        std::list<lru_entry_type>::iterator it;
+    };
+public:
+    using value_type = lw_shared_ptr<cache_entry_type>;
+
+private:
+    struct lru_entry_type {
+        value_type cache_entry;
+        seastar::lowres_clock::time_point timestamp;
+        size_t instance_size;
+    };
+
+private:
+    std::unordered_multimap<cache_key_type, value_type> _cache;
+    std::list<lru_entry_type> _lru;
+    seastar::timer<seastar::lowres_clock> _timer;
+    // The instance in cache time out after up to 2*_timer_period.
+    seastar::lowres_clock::duration _timer_period;
+    // sizes in wasm pages (16KiB)
+    size_t _total_size = 0;
+    size_t _max_size;
+
```

Comment:
Better to count size in bytes to avoid confusion later on.

Reply:
Good idea especially because the page size was incorrect in the comment